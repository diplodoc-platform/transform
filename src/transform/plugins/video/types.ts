import type Token from 'markdown-it/lib/token';
import type {VideoService} from './const';

export type VideoToken = Token & {
    service: string;
    videoID: string;
};

export type VideoServicesOptions = {
    [VideoService.YouTube]: {width: number; height: number};
    [VideoService.Vimeo]: {width: number; height: number};
    [VideoService.Vine]: {width: number; height: number; embed: 'simple' | string};
    [VideoService.Prezi]: {width: number; height: number};
    [VideoService.Osf]: {width: string; height: string};
    [VideoService.Yandex]: {width: number; height: number};
    [VideoService.Vk]: {width: number; height: number};
};

export type VideoFullOptions = VideoServicesOptions & {
    url: VideoUrlFn;
};

export type VideoPluginOptions = Partial<VideoFullOptions>;

export type VideoUrlFn = (
    service: string,
    videoID: string,
    options: VideoServicesOptions,
) => string;
