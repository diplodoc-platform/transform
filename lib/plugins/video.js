const video = require('markdown-it-video');

module.exports = video;
