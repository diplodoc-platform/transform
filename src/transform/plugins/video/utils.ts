import type {VideoUrlFn} from './types';

export const videoUrl: VideoUrlFn = (service, videoID, options) => {
    switch (service) {
        case 'youtube':
            return `https://www.youtube.com/embed/${videoID}`;
        case 'vimeo':
            return `https://player.vimeo.com/video/${videoID}`;
        case 'vine':
            return `https://vine.co/v/${videoID}/embed/${options.vine.embed}`;
        case 'prezi':
            return (
                `https://prezi.com/embed/${videoID}` +
                '/?bgcolor=ffffff&amp;lock_to_path=0&amp;autoplay=0&amp;autohide_ctrls=0&amp;' +
                'landing_data=bHVZZmNaNDBIWnNjdEVENDRhZDFNZGNIUE43MHdLNWpsdFJLb2ZHanI5N1lQVHkxSHFxazZ0UUNCRHloSXZROHh3PT0&amp;' +
                'landing_sign=1kD6c0N6aYpMUS0wxnQjxzSqZlEB8qNFdxtdjYhwSuI'
            );
        case 'osf':
            return `https://mfr.osf.io/render?url=https://osf.io/${videoID}/?action=download`;
        case 'yandex':
            return `https://runtime.video.cloud.yandex.net/player/video/${videoID}`;
        case 'vk':
            return `https://vk.com/video_ext.php?${videoID}`;
        default:
            return service;
    }
};
