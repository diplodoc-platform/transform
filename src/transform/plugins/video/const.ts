import type {VideoFullOptions} from './types';

import {videoUrl} from './utils';

/** Supported services */
export enum VideoService {
    YouTube = 'youtube',
    Vimeo = 'vimeo',
    Vine = 'vine',
    Prezi = 'prezi',
    Osf = 'osf',
    Yandex = 'yandex',
    Vk = 'vk',
}

export const defaults: VideoFullOptions = {
    url: videoUrl,
    youtube: {width: 640, height: 390},
    vimeo: {width: 500, height: 281},
    vine: {width: 600, height: 600, embed: 'simple'},
    prezi: {width: 550, height: 400},
    osf: {width: '100%', height: '100%'},
    yandex: {width: 640, height: 390},
    vk: {width: 640, height: 390},
};
