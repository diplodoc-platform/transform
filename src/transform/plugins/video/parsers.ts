import {VideoService} from './const';

const ytRegex = /^.*((youtu.be\/)|(v\/)|(\/u\/\w\/)|(embed\/)|(watch\?))\??v?=?([^#&?]*).*/;
export function youtubeParser(url: string) {
    const match = url.match(ytRegex);
    return match && match[7].length === 11 ? match[7] : url;
}

const vimeoRegex =
    /https?:\/\/(?:www\.|player\.)?vimeo.com\/(?:channels\/(?:\w+\/)?|groups\/([^/]*)\/videos\/|album\/(\d+)\/video\/|)(\d+)(?:$|\/|\?)/;
export function vimeoParser(url: string) {
    const match = url.match(vimeoRegex);
    return match && typeof match[3] === 'string' ? match[3] : url;
}

const vineRegex = /^http(?:s?):\/\/(?:www\.)?vine\.co\/v\/([a-zA-Z0-9]{1,13}).*/;
export function vineParser(url: string) {
    const match = url.match(vineRegex);
    return match && match[1].length === 11 ? match[1] : url;
}

const preziRegex = /^https:\/\/prezi.com\/(.[^/]+)/;
export function preziParser(url: string) {
    const match = url.match(preziRegex);
    return match ? match[1] : url;
}

// TODO: Write regex for staging and local servers.
const mfrRegex =
    /^http(?:s?):\/\/(?:www\.)?mfr\.osf\.io\/render\?url=http(?:s?):\/\/osf\.io\/([a-zA-Z0-9]{1,5})\/\?action=download/;
export function mfrParser(url: string) {
    const match = url.match(mfrRegex);
    return match ? match[1] : url;
}

const yandexRegex = /^https:\/\/runtime.video.cloud.yandex.net\/player\/video\/([a-zA-Z0-9]+)/;
export function yandexParser(url: string) {
    const match = url.match(yandexRegex);
    return match ? match[1] : url;
}

const vkRegex = /^https:\/\/vk.com\/video_ext\.php?(oid=[-\d]+&id=[-\d]+)/;
export function vkParser(url: string) {
    const match = url.match(vkRegex);
    return match ? match[1] : url;
}

export function parseVideoUrl(service: string, url: string): string | false {
    let videoID = '';

    switch (service.toLowerCase()) {
        case VideoService.YouTube:
            videoID = youtubeParser(url);
            break;
        case VideoService.Vimeo:
            videoID = vimeoParser(url);
            break;
        case VideoService.Vine:
            videoID = vineParser(url);
            break;
        case VideoService.Prezi:
            videoID = preziParser(url);
            break;
        case VideoService.Osf:
            videoID = mfrParser(url);
            break;
        case VideoService.Yandex:
            videoID = yandexParser(url);
            break;
        case VideoService.Vk:
            videoID = vkParser(url);
            break;
        default:
            return false;
    }

    // If the videoID field is empty, regex currently make it the close parenthesis.
    if (videoID === ')') {
        videoID = '';
    }

    return videoID;
}
