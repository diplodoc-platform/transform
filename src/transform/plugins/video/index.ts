// Process @[youtube](youtubeVideoID)
// Process @[vimeo](vimeoVideoID)
// Process @[vine](vineVideoID)
// Process @[prezi](preziID)
// Process @[osf](guid)

import type MarkdownIt from 'markdown-it';
// eslint-disable-next-line no-duplicate-imports
import type {PluginWithOptions} from 'markdown-it';
import type ParserInline from 'markdown-it/lib/parser_inline';
import type Renderer from 'markdown-it/lib/renderer';

import type {VideoFullOptions, VideoPluginOptions, VideoToken} from './types';
import {parseVideoUrl} from './parsers';
import {VideoService, defaults} from './const';

// eslint-disable-next-line valid-jsdoc
/**
 * Video plugin for markdown-it.
 * Forked from https://github.com/CenterForOpenScience/markdown-it-video/tree/0.6.3
 */
const video: PluginWithOptions<VideoPluginOptions> = (md, options) => {
    const theOptions: VideoFullOptions = {...defaults, ...options};
    const theMd = md;

    theMd.renderer.rules.video = tokenizeVideo(theMd, theOptions);
    theMd.inline.ruler.before('emphasis', 'video', videoEmbed(theMd, theOptions));
};

function tokenizeVideo(md: MarkdownIt, options: VideoFullOptions): Renderer.RenderRule {
    return (tokens, idx) => {
        const videoID = md.utils.escapeHtml((tokens[idx] as VideoToken).videoID);
        const service = md.utils.escapeHtml((tokens[idx] as VideoToken).service).toLowerCase();
        const checkUrl =
            /http(?:s?):\/\/(?:www\.)?[a-zA-Z0-9-:.]{1,}\/render(?:\/)?[a-zA-Z0-9.&;?=:%]{1,}url=http(?:s?):\/\/[a-zA-Z0-9 -:.]{1,}\/[a-zA-Z0-9]{1,5}\/\?[a-zA-Z0-9.=:%]{1,}/;
        let num;

        if (service === VideoService.Osf && videoID) {
            num = Math.random() * 0x10000;

            if (videoID.match(checkUrl)) {
                return (
                    '<div id="' +
                    num +
                    '" class="mfr mfr-file"></div><script>' +
                    '$(document).ready(function () {new mfr.Render("' +
                    num +
                    '", "' +
                    videoID +
                    '");' +
                    '    }); </script>'
                );
            }
            return (
                '<div id="' +
                num +
                '" class="mfr mfr-file"></div><script>' +
                '$(document).ready(function () {new mfr.Render("' +
                num +
                '", "https://mfr.osf.io/' +
                'render?url=https://osf.io/' +
                videoID +
                '/?action=download%26mode=render");' +
                '    }); </script>'
            );
        }

        const {width, height} = options[service as VideoService];

        return videoID === ''
            ? ''
            : '<div class="embed-responsive embed-responsive-16by9"><iframe class="embed-responsive-item ' +
                  service +
                  '-player" type="text/html" width="' +
                  width +
                  '" height="' +
                  height +
                  '" src="' +
                  options.url(service, videoID, options) +
                  '" frameborder="0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe></div>';
    };
}

const EMBED_REGEX = /@\[([a-zA-Z].+)]\([\s]*(.*?)[\s]*[)]/im;
// eslint-disable-next-line @typescript-eslint/no-unused-vars
function videoEmbed(md: MarkdownIt, _options: VideoFullOptions): ParserInline.RuleInline {
    return (state, silent) => {
        const theState = state;
        const oldPos = state.pos;

        if (
            state.src.charCodeAt(oldPos) !== 0x40 /* @ */ ||
            state.src.charCodeAt(oldPos + 1) !== 0x5b /* [ */
        ) {
            return false;
        }

        const match = EMBED_REGEX.exec(state.src.slice(state.pos, state.src.length));

        if (!match || match.length < 3) {
            return false;
        }

        const service = match[1];
        const videoID = parseVideoUrl(service, match[2]);

        if (videoID === false) {
            return false;
        }

        const serviceStart = oldPos + 2;
        const serviceEnd = md.helpers.parseLinkLabel(state, oldPos + 1, false);

        //
        // We found the end of the link, and know for a fact it's a valid link;
        // so all that's left to do is to call tokenizer.
        //
        if (!silent) {
            theState.pos = serviceStart;
            // @ts-expect-error
            theState.service = theState.src.slice(serviceStart, serviceEnd);
            const newState = new theState.md.inline.State(service, theState.md, theState.env, []);
            newState.md.inline.tokenize(newState);

            const token = theState.push('video', '', 0);
            (token as VideoToken).videoID = videoID;
            (token as VideoToken).service = service;
            token.level = theState.level;
        }

        theState.pos += theState.src.indexOf(')', theState.pos);
        return true;
    };
}

export = video;
