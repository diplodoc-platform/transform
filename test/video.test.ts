import path from 'path';
import MarkdownIt from 'markdown-it';

import video from '../src/transform/plugins/video/index';

const generate = require('markdown-it-testgen');

function preserveId(html: string) {
    return html
        .replace(/<div id="\d+.\d+"/, '<div id="1.2"')
        .replace(/new mfr\.Render\("\d+.\d+"/, 'new mfr.Render("1.2"');
}

describe('md-video', () => {
    const md = new MarkdownIt({}).use(video);
    generate(path.join(__dirname, 'data/video/video-fixtures.txt'), md);
});

// Because the mfr iframe requires a random id these tests cannont be part of
// the markdown-it-testgen fixture
describe('md-video-mfr', () => {
    const md = new MarkdownIt({}).use(video);

    it('make sure normal iframe generates properly when empty', () => {
        const renderedHtml = preserveId(md.render('@[osf]()'));
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with guid', () => {
        const renderedHtml = preserveId(md.render('@[osf](xxxxx)'));
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with guid and line break', () => {
        const renderedHtml = preserveId(md.render('@[osf](xxxxx\n)'));
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with guid and extra space', () => {
        const renderedHtml = preserveId(md.render('@[osf](xxxxx )'));
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with guid and two extra spaces', () => {
        const renderedHtml = preserveId(md.render('@[osf]( xxxxx )'));
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with link', () => {
        const renderedHtml = preserveId(
            md.render(
                '@[osf](https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render)',
            ),
        );
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with link and extra space', () => {
        const renderedHtml = preserveId(
            md.render(
                '@[osf](https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render )',
            ),
        );
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with link and two extra spaces', () => {
        const renderedHtml = preserveId(
            md.render(
                '@[osf](https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render )',
            ),
        );
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with link to staging', () => {
        const renderedHtml = preserveId(
            md.render(
                '@[osf](https://mfr-staging3.osf.io/render?url=https://staging3.osf.io/xxxxx/?action=download%26mode=render)',
            ),
        );
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with link to local', () => {
        const renderedHtml = preserveId(
            md.render(
                '@[osf](https://localhost:7778/render?url=https://localhost:5000/xxxxx/?action=download%26mode=render)',
            ),
        );
        expect(renderedHtml).toMatchSnapshot();
    });

    it('make sure normal iframe generates properly with link to local ip', () => {
        const renderedHtml = preserveId(
            md.render(
                '@[osf](http://localhost:7778/render?mode=render&url=http://192.168.168.167:5000/y98tn/?action=download%26mode=render%26direct)',
            ),
        );

        expect(renderedHtml).toMatchSnapshot();
    });
});
