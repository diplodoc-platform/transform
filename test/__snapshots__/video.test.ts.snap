// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`md-video-mfr make sure normal iframe generates properly when empty 1`] = `
<p>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with guid 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with guid and extra space 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with guid and line break 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with guid and two extra spaces 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with link 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with link and extra space 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with link and two extra spaces 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr.osf.io/render?url=https://osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with link to local 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://localhost:7778/render?url=https://localhost:5000/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with link to local ip 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "http://localhost:7778/render?mode=render&amp;url=http://192.168.168.167:5000/y98tn/?action=download%26mode=render%26direct");    });
  </script>
</p>
`;

exports[`md-video-mfr make sure normal iframe generates properly with link to staging 1`] = `
<p>
  <div id="1.2"
       class="mfr mfr-file"
  >
  </div>
  <script>
    $(document).ready(function () {new mfr.Render("1.2", "https://mfr-staging3.osf.io/render?url=https://staging3.osf.io/xxxxx/?action=download%26mode=render");    });
  </script>
</p>
`;
